### ISSUE TITLE
Add poly-A trimming

### ISSUE THREAD

**Issue by marcelm** [marcelm/cutadapt#696]
Inspired by the discussion in #694.

We currently recommend to use `-a A{100}` for trimming poly-A tails, but this a bit of a workaround. There are two issues:
- Too few bases may be trimmed if the poly-A tail is longer than 100 bases
- 100 bases allows 10 errors by default, which means that the first couple of bases are often mismatched to bases in the non-poly-A tail part of the read.
- This uses the full adapter alignment algorithm, which makes this slower than it could be.

To solve the first issue, one could increase the length, but then the second issue gets worse.

An alternative is to add a special option for trimming poly-A tails, which could be faster and more accurate.

**Comment by rhpvorderman** [comment]
With this algorithm what will be the sequence for?

```
GGGGGGGGAAAGAAGAAGAAGAAGAAGAAGAAA
```
I think it will be `GGGGGGGG`, given that only 2 points are deducted for a non-A base.

**Comment by marcelm** [comment]
Yes, that’s the result. Essentially, the algorithm assumes that there is a poly-A tail in each read, which is probably not how it should be done. Let me see.

**Comment by marcelm** [comment]
If the mismatch penalty is increased to 3, the example above would need to be modified to look like this:
```
GGGGGGGGAAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAA
```

Let’s say `T` is a mismatch and `G` describe the sequence without the poly-A tail. Then mismatch penalty 3 would trim this
```
GGGGAAATAAAAAAAAAA
```
to
```
GGGGAAAT
```
Even higher mismatch penalties make that unrecognized stretch of `A` between the last `G` and the `T` longer and longer.

Increasing the mismatch penalty to 9 would be somewhat equivalent to allowing an error rate of 0.1, but then that stretch becomes too long.

I initially had another criterion in the code that counted the errors, maybe I should re-add that.

**Comment by marcelm** [comment]
I added a hard-coded error rate of 0.2 for now (and left the mismatch penalty at 2), which fixes your example in the sense that the sequence is now trimmed to
```
GGGGGGGGAAAGAAGAAGAAGAAGAAGAAG
```
(only the `AAA` suffix is removed)

I declared this as "experimenal" in the changelog, so this can be tuned further (and I’m sure there are still edge cases). I’ll try to find someone who can test this on real-world data.

**Comment by rhpvorderman** [comment]
This is much better. What also would work is a sliding window of certain size (say 8). Which slides from the end to the beginning of the sequence. In the sliding window a maximum of 1 error is allowed. As soon as more than 1 error is encountered, the sequence is cut off at the latest A. Using a sliding window prevents overcutting because previous stretches of unaltered A are "forgotten". But a sliding window can be complex and computing it the naive way requires length of polyA tail times sliding window size calculations, so that is not ideal to say the least.

**Comment by y9c** [comment]
This is a very useful argument, but it would be better if the read orientation is taken in to account. 

Example 1: (Real polyA)

```
==> test1.fa <==
>test /R1
GAGAGACAGAGACCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

==> test2.fa <==
>test /R2
TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGGGGTCTCTGTCTCTC
```

After cutadapt with `--poly-a`, 

```
>test /R1
GAGAGACAGAGACCCC
>test /R2
TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGGGGTCTCTGTCTCTC
```

Expect output would be:

```
>test /R1
GAGAGACAGAGACCCC
>test /R2
GGGGTCTCTGTCTCTC
```

---

Example 2: (not polyA)

```
==> test1.fa <==
>test /R1
GAGAGACAGAGACCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

==> test3.fa <==
>test /R2
GTCTCTGTCTGGGACTGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```

After `--poly-a`

```
>test /R1
GAGAGACAGAGACCCC
>test /R2
GTCTCTGTCTGGGACTG
```

**Comment by marcelm** [comment]
Hi, I agree this needs to be adjusted so that it is usable for paired-end reads. Please remind me in one week if I haven’t gotten back to this by then.

**Comment by y9c** [comment]
Thank you @marcelm.

I would like to say that it might be a little complicated, since RNA seq library can be either first-stranded or second-stranded. Do you think we need an another argument to toggle this? Or let the tool detect automatically?

**Comment by nolanmaier** [comment]
Hi, I have been using the `--poly-a` flag on single-end reads and it has been working great. However, this flag does not output any statistics into the report, so it is not easy to tell how many reads were trimmed. One suggestion is to treat the `--poly-a` flag as an adapter and output per-adapter statistics as normal. This would more closely match how trimming of poly-a sequences was reported using the old `-a A{100}` method.

**Comment by marcelm** [comment]
Good point, this would be helpful. I won’t have time for this over the next weeks due to vacation etc., but feel free to remind me in August.

**Comment by dlaehnemann** [comment]
Many thanks for adding this option, it is very useful to have something more specific than `-a A{100}` that automatically gets applied at the correct stage (after adapter removal!).

Just a +1 from me for some statistics output in the report. And a more detailed description in the docs of what the algorithm behind the option actually does, would also be great.

**Comment by marcelm** [comment]
I have just added a section to the documentation [describing the poly-A trimming algorithm](https://cutadapt.readthedocs.io/en/latest/algorithms.html#poly-a-algorithm).

Regarding the statistics, note that the number of trimmed basepairs is already reported:
```
=== Summary ===

Total reads processed:                       3
Reads written (passing filters):             3 (100.0%)

Total basepairs processed:           148 bp
Poly-A-trimmed:                       84 bp (56.8%)   <-------here
Total written (filtered):             64 bp (43.2%)
```

I understand that a length histogram is also interesting, so I added that just now as well. It looks like this:
```
=== Poly-A trimmed ===

length  count
21      2
42      1
```

This looks similar to the length histogram for trimmed adapters, but with two columns only (it didn’t make sense to include the "expect" column, and the number of errors is not available). This means that MultiQC won’t be able to parse this at the moment.

@y9c and @MurakawaLab I’d like to make this work for paired-end data. Could you point me to some example data with poly-A tails? (That is, with poly-A on R1 and poly-T on R2.)

**Comment by dlaehnemann** [comment]
Ah yes, I remember seeing this in the log files during testing of the `--poly-a` command line flag. But funnily, with the following combination of command line arguments, this does not show up in the logs:

```
This is cutadapt 4.4 with Python 3.10.12
Command line parameters: --cores 8 -a r1adapter=AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC;min_overlap=7;max_error_rate=0.005 --minimum-length 33 --nextseq-trim=20 --poly-a -o path/to/out.fastq.gz path/to/in/single_end.fastq.gz
Processing single-end reads on 8 cores ...
Finished in 66.466 s (5.585 Âµs/read; 10.74 M reads/minute).

=== Summary ===

Total reads processed:              11,901,529
Reads with adapters:                 1,680,520 (14.1%)

== Read fate breakdown ==
Reads that were too short:             461,240 (3.9%)
Reads written (passing filters):    11,440,289 (96.1%)

Total basepairs processed: 1,202,054,429 bp
Quality-trimmed:              10,535,736 bp (0.9%)
Total written (filtered):  1,035,645,758 bp (86.2%)

=== Adapter r1adapter ===
```

Or is there anything I am doing wrong here? Any incompatible command line options or so? Or is this some funny (buggy?) interaction of some arguments?

**Comment by marcelm** [comment]
> Or is there anything I am doing wrong here? Any incompatible command line options or so? Or is this some funny (buggy?) interaction of some arguments?

It was the latter, thanks for the report! This was an issue when running on multiple cores (in that case, the statistics from each thread need to be added up to get the final statistics and the code for doing that was missing). Should be fixed now.

**Comment by marcelm** [comment]
I’ve now also implemented proper poly-A trimming for paired-end data: With `--poly-a`, poly-A tails are removed from R1 and "poly-T heads" are removed from R2. I think this covers everything mentioned in this issue. Please open a new one for further problems or feature requests.

Any volunteer who could test this before the release?
